Feature/initial dev anonaka

anonaka opened the pull request:
my first version of code ready for review.

Commit:
new today

Commit:
WIP: frist wroking version

anonaka committed:
added get_delivery_status. needs refactor.

anonaka committed:
WIP, changed to OO api

anonaka committed:
add more files

anonaka committed:
get status updated

anonaka committed:
packaging stuff

anonaka committed:
fix typo.

anonaka committed:
added voice api

anonaka committed:
remove json

almond55 commented:
Maybe just `playback()`?

anonaka commented:
How about "send_voice" to be symmetric with send_sms?

k4ml commented:
Let's take this discussion to Slack instead.

anonaka committed:
added import

k4ml commented on xoxzo/cloudpy/cloudpy.py:
```diff
@@ -0,0 +1,47 @@
+        self.voice_palyback_last_respose = None
+
+    def send_sms(self, message, recipient, sender):
+        url = "https://api.xoxzo.com/sms/messages/"
```
You can define a BASE_URL as a class variable, and then also make it possible to override it using environment variables. That make it possible for us to point the client to alternative endpoint, such as our development/local endpoint.

Also I think the sid and auth_token value you have to reference it from environment variables to avoid hardcoding it in the code itself.

anonaka committed:
get some parameters from env
first unit test code

anonaka committed:
disable some test cases

anonaka committed:
rename to tests

Commit:
Update initial README

Commit:
Add replacement README

Commit:
Merge pull request #2 from iq8al/master

Removed old README and replace it with a new one

anonaka committed:
Merge remote-tracking branch 'origin' into feature/initial-dev-anonaka

anonaka committed:
[WIP]
rename voice	method.
added more tests.
start documentation.

anonaka committed:
better error handling
added more tests.

anonaka committed:
added error check for call  simple playback.
added tests.

anonaka committed:
WIP, test

anonaka committed:
WIP, added sample code.
minor fix for tests.

anonaka committed:
added sample code.

anonaka committed:
added more documentation.

anonaka committed:
WIP,
add License
do not remember last msgid/callid
doctoring comment
clean up tests.

anonaka committed:
updated.

anonaka committed:
MIT license.

anonaka committed:
ignore build dir

anonaka committed:
clear pep8 check.

anonaka committed:
set requests version.

anonaka committed:
added build dir.

anonaka committed:
call play back sample code added.

anonaka committed:
added explanation sample2.

anonaka committed:
add assertion for each test.

anonaka committed:
removed unnecessary line.

anonaka committed:
docstring are updated.

k4ml commented on xoxzo/cloudpy/cloudpy.py:
```diff
@@ -0,0 +1,126 @@
+    :param string auth_token: Your auth_token of the xoxzo account.
+        If None, value of the environment variable XOXZO_API_AUTH_TOKEN
+        will be used.
+    '''
```
@anonaka Can you remove getting the credentials from env var ? It's an oversight of mine. Our library shouldn't really concern how user will get the credentials. It should use what being passed to `__init__()` as is. We should look into env var if we also provide a CLI command for the library.

I'm sorry for the back and forth.

anonaka replied:
OK!

anonaka committed:
stop getting credentials from env.

k4ml commented on xoxzo/cloudpy/cloudpy.py:
```diff
@@ -21,17 +21,15 @@ class XoxzoClient:
+        self.auth_token = auth_token
         # you can override api host by setting envrionment
         # variable XOXZO_API_HOST
         api_host = os.environ.get("XOXZO_API_HOST")
```
Please remove this one too. You can add a third optional parameter to `__init__()` for `api_host`.

anonaka replied:
fixed.

anonaka committed:
stop using  XOXZO_API_HOST.

anonaka committed:
add get_sent_sms_list().

anonaka committed:
updated method name.

anonaka committed:
WIP use Xoxzo response object.

anonaka committed:
temporary WIP push to move to PyCharm.

anonaka committed:
Fixed tests.

anonaka committed:
Catch requests exception.

anonaka committed:
Catch requests exception.

anonaka committed:
WIP, fixed the bug, set the response in the correct field.

anonaka committed:
added more assertions in the tests.
return dict in case exception thrown.

k4ml commented on xoxzo/cloudpy/cloudpy.py:
```diff
@@ -0,0 +1,199 @@
+            self.xoxzo_api_sms_url,
+            data=payload,
+            auth=(self.sid, self.auth_token))
+            if req_res.status_code == 201:
```
You should check this outside of try block. Try block should be as simple as possible:-

``` python
try:
    resp = requests.post(.....)
except Exception as e:
    # handle exception

...
# the rest of the logic
```

k4ml commented on xoxzo/cloudpy/cloudpy.py:
```diff
@@ -0,0 +1,199 @@
+    :param string auth_token: Your auth_token of the xoxzo account.
+    :param string api_host: If None, stadard xoxzo server will be used.
+    '''
+    REQUESTS_EXCEPITON = 499
```
What this REQUESTS_EXCEPTION ?

anonaka replied:
This is the value set in XoxzoResponse.errors when cloudpy  catches the Requests exception.

k4ml commented on xoxzo/cloudpy/cloudpy.py:
```diff
@@ -0,0 +1,199 @@
+        except requests.exceptions.RequestException as e:
+            xr =  XoxzoResponse(errors=XoxzoClient.REQUESTS_EXCEPITON, message= {"http_error":e})
+        finally:
+            return xr
```
Using finally here will lead to subtle bug, as it will execute regardless if there's an error in the try and except block above. The rule of thumb is simple, we use `finally` when we can't find any other way anymore. So far I've only need to use finally in some very rare cases.

k4ml commented:
@anonaka As I've commented in one of the changeset, I noticed lot of logic inside try block in all of the functions. Please fix it.

anonaka committed:
Made try block smaller.
Added some documentation.

anonaka committed:
minor doc fix.
